syntax = "proto2";

package arf_proto;

import "Time.proto";
import "UUID.proto";

message DataEvent {
    enum Type {
        CREATION = 0;
        MOVEMENT = 1;
        DESTRUCTION = 2;
    }

    required Type type = 1;
    required UUID data_id = 2;
    repeated UUID parent_ids = 3;
    required Time mono_time = 4;
    required Time wall_time = 5;
}

message TaskEvent {
    enum Type {
        PENDING = 0;
        CANCELLED = 1;
        STARTING = 2;
        RUNNING = 3;
        FINISHED = 4;
        ERRORED = 5;
    }

    required Type type = 1; 
    required UUID task_id = 2;
    repeated UUID parent_ids = 3;
    required Time mono_time = 4;
    required Time wall_time = 5;
}

message InitEvent {
    enum Type {
        CREATE_NODE = 0;
        CREATE_INPUT_PORT = 1;
        CREATE_OUTPUT_PORT = 2;
        CREATE_WORKER_THREAD = 3;
        CONNECT_PORTS = 4;
    }

    required Type type = 1;
    optional UUID object_id = 2;
    optional string object_name = 3;
    optional UUID parent_id = 4;
    optional string data_type = 5;
    optional UUID output_port_id = 6;
    optional UUID input_port_id = 7;
    required Time mono_time = 8;
    required Time wall_time = 9;
}

message ArfEventItem {
    oneof entry {
        DataEvent data_event = 1;
        TaskEvent task_event = 2;
        InitEvent init_event = 3;
    }
}